lib/Target/SubstraitPB/Export.cpp

```diff
@@ -0,0 +1,191 @@
+//===-- Export.cpp - Export Substrait dialect to protobuf -------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "structured/Target/SubstraitPB/Export.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Support/LogicalResult.h"
+#include "structured/Dialect/Substrait/IR/Substrait.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#include <google/protobuf/text_format.h>
+#include <substrait/algebra.pb.h>
+#include <substrait/plan.pb.h>
+#include <substrait/type.pb.h>
+
+using namespace mlir;
+using namespace mlir::substrait;
+
+namespace pb = google::protobuf;
+
+namespace {
+
+// Forward declaration for the export function of the given operation type.
+//
+// We need one such function for most op type that we want to export. The
+// forward declarations are necessary such all export functions are available
+// for the definitions indepedently of the order of these definitions. The
+// `MESSAGE_TYPE` argument corresponds to the protobuf message type returned
+// by the function.
+#define DECLARE_EXPORT_FUNC(OP_TYPE, MESSAGE_TYPE)                             \
+  static FailureOr<std::unique_ptr<::substrait::MESSAGE_TYPE>>                 \
+  exportOperation(OP_TYPE op);
+
+DECLARE_EXPORT_FUNC(ModuleOp, Plan)
+DECLARE_EXPORT_FUNC(NamedTableOp, Rel)
+DECLARE_EXPORT_FUNC(PlanOp, Plan)
+DECLARE_EXPORT_FUNC(RelOpInterface, Rel)
+
+FailureOr<std::unique_ptr<::substrait::Type>> exportType(Type mlirType) {
+  // TODO(ingomueller): Support other types.
+  auto si32 = IntegerType::get(mlirType.getContext(), 32, IntegerType::Signed);
+  if (mlirType != si32)
+    return failure();
+
+  // TODO(ingomueller): support other nullability modes.
+  auto i32Type = std::make_unique<::substrait::Type::I32>();
+  i32Type->set_nullability(
+      ::substrait::Type_Nullability::Type_Nullability_NULLABILITY_REQUIRED);
+
+  auto type = std::make_unique<::substrait::Type>();
+  type->set_allocated_i32(i32Type.release());
+
+  return std::move(type);
+}
+
+FailureOr<std::unique_ptr<::substrait::Plan>> exportOperation(ModuleOp op) {
+  if (!op->getAttrs().empty())
+    return failure();
+
+  if (op->getRegions().size() != 1)
+    return failure();
+
+  Region &body = op.getBodyRegion();
+  if (llvm::range_size(body.getOps()) != 1)
```

Comment:
This is a little wasteful/one could early exit without computing total size.

Reply:
How? The only way I can think of is something like `++body.getOps().begin() != body.op_end()` but that's much less readable, isn't it?

Reply:
Indeed it is, one could also if the first is a terminator, I think your version from comment would be cheaper than that though. One could get creative with zip and seq too , but I think simplest may be a lambda with loop inside from readability point of view (I mean this may not matter, I meant "little" in the non-British use of little :-))

Reply:
Shame! [`std::ranges::take_view`](https://en.cppreference.com/w/cpp/ranges/take_view) would be helpful here but it's C++20.

But, actually, maybe `std::next(body.op_begin()) != body.op_end()` isn't too bad?

Reply:
Yet another thought: if the IR is valid, then the range of the size is 1, so `llvm::range_size` acutally has acceptable cost. Do we really have to optimize for runtime on invalid IR?

Reply:
Based on that last argument, I propose to revert my latest change and go with the initial version. WDYT?

Reply:
Resolving this discussion because (1) the related line wasn't part of this commit, (2) it's the only one left for an otherwise approved PR, (3) we can always come back later: #815.